### ISSUE TITLE
[BUG]:  Model::find() protected/private property access public

### ISSUE THREAD

**Issue by tidytrax** [phalcon/cphalcon#14766]
Using get-set and protected property

```
class Model extends \Phalcon\Mvc\Model
{
    protected $prop;

    public function getProp()
    {
        return $this->prop;
    }

    public function setProp($prop)
    {
        $this->prop = $prop;
        return $this;
    }
}
```

```
$model = new Model();
**_Protected, Can't access._**
echo $model->prop;
$model->prop = 1;

**_Work can access_**
$model->setProp() OR $model->getProp(); 
```

```
$resultset = Model::find();
foreach ( $resultset as $row ) {
   echo $row->prop; //Works
   $row->prop = 1;
}
```

Config
Phalcon 4.0.2 / PHP 7.4.2 / Postgresql 12.1/Nginx 1.16.1

**Comment by ruudboon** [comment]
@tidytrax Is this occurring with properties as private as well?

**Comment by tidytrax** [comment]
> @tidytrax Is this occurring with properties as private as well?

Yes I tested now and **private work** also, same code above, but with private prop

foreach ( Model::find() as $row ) {
   var_dump($row->prop);
}

**Comment by ruudboon** [comment]
@tidytrax thnx. We will have a look at it

**Comment by tidytrax** [comment]
> @tidytrax thnx. We will have a look at it

Thanks, it's not a big problem, but the postgres boolean([14722](https://github.com/phalcon/cphalcon/issues/14722)) is killing my app. 

Could you provide me with some turn around ? like someway to bring all fields as string.

**Comment by tidytrax** [comment]
The bug occurs on Model::findFirst() also

**Comment by ruudboon** [comment]
I wasn't able to reproduce this. Can you check the type of object you get from the result set?
Maybe the hydrate mode is incorrect?
```
$resultset->setHydrateMode(
            Resultset::HYDRATE_RECORDS
);
```

**Comment by tidytrax** [comment]
they came with boolean type if i do a $resultset->toArray();
They came correct, they do not appear on changed fields.

Same result as before.
```
$model = UserSession::findFirst(['hydration' => Resultset::HYDRATE_RECORDS]);
var_dump($model->toArray());
```

> id: (integer) 2 
> url: (string) "/logout"
> ip: (string) "10.0.2.2"
> force_logout: (boolean) false 
> remember: (boolean) false 
> 

```
$model->id = 1;
$model->save();
var_dump($model->getMessages());
```

> Phalcon\MessagesMessage (object) [Object ID #124][5 properties]
> code:protected: (integer) 0 
> field:protected: (string) "force_logout"
> message:protected: (string) "force_logout is required"
> type:protected: (string) "PresenceOf"
> metaData:protected: 
> (array) [0 elements]
> 1: 
> Phalcon\MessagesMessage (object) [Object ID #123][5 properties]
> code:protected: (integer) 0 
> field:protected: (string) "remember"
> message:protected: (string) "remember is required"
> type:protected: (string) "PresenceOf"
> metaData:protected: 
> (array) [0 elements]



_**I didn't updated this 2 fields, they are filled, but ask again, and if i try**_
```
$model->force_logout = false/true: //**fail**
$model->force_logout = 0/1 //**works**.
```

**This is fine.**
`var_dump($model->getChangedFields());exit;`

> (array) [2 elements]
> 0: (string) "session_id"
> 1: (string) "updated"
> 

So always when i want to update a record i have to pass all boolean columns as int again, even if they are already filledup, **phalcon does not accept to save bool,**

**How i have to do to update a record on database or save.**

```
$model = UserSession::findFirst(['hydration' => Resultset::HYDRATE_RECORDS]);
$model->id = 1;
$model->force_logout = ( int) $model->force_logout(already has a value);
$model->remember = (int) $model->remember(already has a value);
$model->save();
```

**Comment by tidytrax** [comment]
```
$model = UserSession::findFirst(['hydration' => Resultset::HYDRATE_RECORDS]);
$model->url = '/';
$model->save();
```

Work's even with hydrate.

**Comment by tidytrax** [comment]
Just a info, The Model extends an AbstractModel 
The Abstract Model has the proprerties
And the public get-set

**Comment by ruudboon** [comment]
> Just a info, The Model extends an AbstractModel
Like this? 

```
class AbstractModel extends \Phalcon\Mvc\Model
{
    protected $prop;

    public function getProp()
    {
        return $this->prop;
    }

    public function setProp($prop)
    {
        $this->prop = $prop;
        return $this;
    }
}

class MyModel extends AbstractModel {

}

$model = MyModel::findFirst();
$model->prop = 'secret';
$model->save();
```

**Comment by tidytrax** [comment]
**Yes exactly .**
works on both way $model->prop or $model->setProp()

**Comment by ruudboon** [comment]
Will write some extra tests for this to see if we can reproduce this. Are you on Linux?

**Comment by tidytrax** [comment]
> Will write some extra tests for this to see if we can reproduce this. Are you on Linux?

Vagrant 2.2.6 VirtualBox 6.0.14 Ubuntu 19.10
Host Windows 10.

**Comment by Jurigag** [comment]
Related https://github.com/phalcon/zephir/pull/2056 https://github.com/phalcon/zephir/issues/2057

**Comment by sergeyklay** [comment]
Zephir issue has been fixed in #2078. I'll release new version shortly

**Comment by sergeyklay** [comment]
Zephir 0.12.18 released: https://github.com/phalcon/zephir/releases/tag/0.12.18
